File: archive/VGM_functions/import_odes.m
%% Import ODEs
% Authors: Nico Stephan Gorbach and Stefan Bauer
function [ode,coupling_idx] = import_odes(symbols,odes_path)

%%
% Import ODEs expressions
ode.system_string = importdata(odes_path);

%%
% Refine ODEs
assume(symbols.state,'real'); assume(symbols.param,'real');
for k = 1:length(ode.system_string)
    for u = 1:length(symbols.state)
        ode.system_string{k} = strrep(ode.system_string{k},...
            ['[' char(symbols.state(u)) ']'],['state(:,' num2str(u) ')']);
    end
    for j = 1:length(symbols.param)
        ode.system_string{k} = strrep(ode.system_string{k},...
            ['[' char(symbols.param(j)) ']'],['param(' num2str(j) ')']);
    end
end
ode.system = cellfun(@(x) str2func(['@(state,param)(' x ')']),ode.system_string,...
    'UniformOutput',false);

ode.system_sym_unpacked = cell2sym(cellfun(@(x) x(symbols.state,symbols.param),ode.system,...
    'UniformOutput',false));

%%
% Packing
param_sym = sym('param%d',[1,length(symbols.param)]); assume(param_sym,'real');
state_sym = sym('state%d',[1,length(symbols.state)]); assume(state_sym,'real');
ode.system_sym = cell2sym(cellfun(@(x) x(state_sym,param_sym),ode.system,'UniformOutput',false));

coupling_idx = find_state_couplings_in_odes(ode);

disp(' '); disp('ODEs:'); disp(' '); 
state_idx = cellfun(@(x) ~strcmp(x(1),'u'),symbols.state_string);
d = sym('d'); dt = sym('dt'); assume(symbols.state,'real');
assume(d,'real'); assume(dt,'real');
pretty(d * symbols.state(state_idx)'/dt == ode.system_sym_unpacked);

end

%% Find state couplings in ODEs
% Authors: Nico Stephan Gorbach and Stefan Bauer

function coupling_idx = find_state_couplings_in_odes(ode)

state_sym = sym('state%d',[1,length(ode.system)]); assume(state_sym,'real');
for k = 1:length(ode.system)
    tmp_idx = ismember(state_sym,symvar(ode.system_sym(k))); tmp_idx(:,k) = 1;
    ode_couplings_states(k,tmp_idx) = 1;
end

for u = 1:size(ode_couplings_states,1)
    coupling_idx.states{u} = find(ode_couplings_states(:,u));
end
end

File: VGM_modules/import_odes.py

# # Proxies for ODE Parameters and States
# 
# #### Authors: Nico S. Gorbach and Stefan Bauer

# In[1]:

# ## Import Numpy and Symbolic Computation Modules


import numpy as np
import sympy as sym


# In[2]:


def import_odes(symbols,odes_path):
    
    '''Imports ODEs from file specified by odes_path'''
    
    with open(odes_path) as f:
        odes_string = f.read().splitlines()
        
    odes_sym = [o.factor() for o in sym.sympify(odes_string)]
    
    # unpack state and parameter symbols
    symbols_all = [lst for sublist in [symbols.state,symbols.param] for lst in sublist]
    odes = sym.lambdify( symbols_all,odes_sym)
    
    return odes,odes_sym

def gradient_of_odes(symbols,odes_sym):
    
    '''Computes gradient of ODEs'''

    symbols_all = [lst for sublist in [symbols.param,symbols.state] for lst in sublist]
    symbols_all.append(sym.symbols('one_vector'))
    
    
    odes_diff_param_sym = [[o.diff(s) for s in symbols.param] for o in odes_sym]
    for j in range(len(odes_diff_param_sym)):
        for i in range(len(odes_diff_param_sym[j])):
            #if len(odes_diff_param_sym[j][i].free_symbols) == 0: odes_diff_param_sym[j][i] = sym.symbols('one_vector') 
            odes_diff_param_sym[j][i] *= sym.symbols('one_vector')
            if odes_diff_param_sym[j][i] == 0: odes_diff_param_sym[j][i] = np.finfo(float).eps * sym.symbols('one_vector') 
    odes_diff_param = sym.lambdify(symbols_all,odes_diff_param_sym)
            
    
    odes_diff_states_sym = [[] for i in range(len(symbols.state))]
    for u in range(len(symbols.state)):
        odes_diff_states_sym[u] = [o.diff(symbols.state[u]) for o in odes_sym]
    
    #odes_diff_states_sym = [[o.expand().diff(s) for s in symbols.state] for o in odes_sym]
    for j in range(len(odes_diff_states_sym)):
        for i in range(len(odes_diff_states_sym[j])):
            odes_diff_states_sym[j][i] *= sym.symbols('one_vector')
            if odes_diff_states_sym[j][i] == 0:
                odes_diff_states_sym[j][i] = np.finfo(float).eps * sym.symbols('one_vector')    
            
     
    odes_diff_states=[[] for u in range(len(symbols.state))]
    for u in range(len(symbols.state)):
        odes_diff_states[u] = sym.lambdify(symbols_all,odes_diff_states_sym[u])
    
    
    return odes_diff_param, odes_diff_states

# In[3]:



def find_state_couplings_in_odes(odes_sym,symbols):
    
    '''Find Couplings  between States across ODEs'''
    
    odes_couplings_to_states = []
    symbols_all = [lst for sublist in [symbols.state,symbols.param] for lst in sublist]
    for u in range(len(symbols.state)):

        
        membership = ismember(symbols.state,odes_sym[u].free_symbols)
        odes_couplings_to_states.append([symbols.state[i] for i,item in enumerate(membership) if item is not None])
    
    state_couplings = [[i for i,item in enumerate(odes_couplings_to_states) if s in item] for s in symbols.state]   
        
    return state_couplings,odes_couplings_to_states


# In[4]:


def ismember(a, b):
    bind = {}
    for i, elt in enumerate(b):
        if elt not in bind:
            bind[elt] = 1
    return [bind.get(itm, None) for itm in a]